templates/vdc/vdc.py

```diff
@@ -170,6 +178,64 @@ def disable(self):
         space.disable('The space should be disabled.')
         self.data['disabled'] = True
 
+    def portforward_create(
+                            self, 
+                            machineId=None, 
+                            port_forwards=[], 
+                            protocol='tcp'):
+        ovc = self.ovc
+        space = self.space        
+        space.get_space_ip()
+        existent_ports = [port['publicPort'] for port in 
+                          ovc.api.cloudapi.portforwarding.list(
+                              cloudspaceId=space.id, machineId=machineId,
+                            )]
+        # add ports
+        for port in port_forwards:
+            # check if ports do not exist yet
+            if str(port['source']) not in existent_ports:
+                # create portforward
+                ovc.api.cloudapi.portforwarding.create(
+                    cloudspaceId=space.id, 
+                    protocol=protocol, 
+                    localPort=port['destination'], 
+                    publicPort=port['source'], 
+                    publicIp=space.get_space_ip(),
+                    machineId=machineId,
+                )
+            else:
+                self.logger.debug("Port forward %s:%s already exists" % (port['destination'], port['source']))
+
+    def portforward_delete(
+                            self, 
+                            machineId=None, 
+                            port_forwards=[], 
+                            protocol='tcp'):
+        ovc = self.ovc
+        space = self.space     
+        existent_ports = [(port['publicPort'], port['localPort'], port['id']) 
+                            for port in ovc.api.cloudapi.portforwarding.list(
+                                            cloudspaceId=space.id, machineId=machineId,
+                                                )]
+        # add ports
+        for publicPort, localPort, id in existent_ports:
+            for port in port_forwards:
+                if publicPort == str(port['source']) and localPort == str(port['destination']):
+                    if str(port['source']) == '22' :
+                        self.logger.error("SSH port 22 can't be deleted")
```

Comment:
Where is this rule comes from ? there is not reason to prevent removing a port forward.
Also, when it's an error, you should raise and not just log